### PR TITLE
Clarify behaviour of developer exception page

### PR DESCRIPTION
**Pull request by jez9999**
Fixes https://github.com/dotnet/aspnetcore/issues/43870

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
That doesn't belong in `:::moniker range=">= aspnetcore-6.0"`, that is, .NET 6 and higher.

**Comment by jez9999** [comment]
Why was this closed @Rick-Anderson ?  I was told elsewhere this docs could do with clarification.

**Comment by Rick-Anderson** [comment]
> `Configure` method.

`Configure` is not used in .NET 6

**Comment by Rick-Anderson** [comment]
> Why was this closed @Rick-Anderson ? I was told elsewhere this docs could do with clarification.

Can you share that?

**Comment by jez9999** [comment]
> > Why was this closed @Rick-Anderson ? I was told elsewhere this docs could do with clarification.
> 
> Can you share that?

https://github.com/dotnet/aspnetcore/issues/43870#issuecomment-1242535767

`IHostBuilder.CreateDefaultBuilder` is still supported in .NET 6, isn't it?  I've read that `WebApplication.CreateBuilder` is actually built on top of it.

**Comment by halter73** [comment]
> `IHostBuilder.CreateDefaultBuilder` is still supported in .NET 6, isn't it?  I've read that `WebApplication.CreateBuilder` is actually built on top of it.

This is correct in .NET 6. In .NET 7, `WebApplication.CreateBuilder` is now built on the new [Host.CreateApplicationBuilder](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.host.createapplicationbuilder?view=dotnet-plat-ext-7.0#microsoft-extensions-hosting-host-createapplicationbuilder) API. It shares a lot of internal logic, but similar to `WebApplicationBuilder`, `HostApplicationBuilder` is less reliant on callbacks. Basically, `IWebHostBuilder` is to `IHostBuilder` what `WebApplicationBuilder` is to `HostApplicationBuilder`.

`IWebHostBuilder` and `IHostBuilder` are still fully supported though. It's not what we use now in our templates and non-hosting-related docs, but this change makes sense to me. This APIs still exist so it's worth calling out that ASP.NET Core doesn't *always* enable the developer exception page by default in development environments.

@DamianEdwards @davidfowl @Tratcher Do you have any opinions on whether we should start enabling the developer exception page by default in development with the generic web host? It'd be easier to explain.

**Review by halter73 — APPROVED** [review]

File: aspnetcore/fundamentals/error-handling.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ This article covers common approaches to handling errors in ASP.NET Core web app
 
 ## Developer exception page
 
-The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when running in the [Development environment](xref:fundamentals/environments).
+The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when running in the [Development environment](xref:fundamentals/environments) if they're created using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder?view=aspnetcore-6.0).  Apps created using the `IHostBuilder.CreateDefaultBuilder` will still need to enable the developer exception page by calling `app.UseDeveloperExceptionPage()` during the startup class's `Configure` method.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:21]:
```suggestion
The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when both:

* Running in the [Development environment](xref:fundamentals/environments).
* App was created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).  Apps created using the [`WebHost.CreateDefaultBuilder`](xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder() must enable the developer exception page by calling `app.UseDeveloperExceptionPage` in `Configure`.
```

File: aspnetcore/fundamentals/error-handling.md [hunk 2]
```diff
@@ -18,7 +18,10 @@ This article covers common approaches to handling errors in ASP.NET Core web app
 
 ## Developer exception page
 
-The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when running in the [Development environment](xref:fundamentals/environments).
+The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when both:
+
+* Running in the [Development environment](xref:fundamentals/environments).
+* App was created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).  Apps created using the [`WebHost.CreateDefaultBuilder`](xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder() must enable the developer exception page by calling `app.UseDeveloperExceptionPage` in `Configure`.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:24]:
```suggestion
* App created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).  Apps created using the [`WebHost.CreateDefaultBuilder`](xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder() must enable the developer exception page by calling `app.UseDeveloperExceptionPage` in `Configure`.
```

File: aspnetcore/fundamentals/error-handling.md [hunk 3]
```diff
@@ -18,7 +18,10 @@ This article covers common approaches to handling errors in ASP.NET Core web app
 
 ## Developer exception page
 
-The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when running in the [Development environment](xref:fundamentals/environments).
+The *Developer Exception Page* displays detailed information about unhandled request exceptions. ASP.NET Core apps enable the developer exception page by default when both:
+
+* Running in the [Development environment](xref:fundamentals/environments).
+* App created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).  Apps created using the [`WebHost.CreateDefaultBuilder`](xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder() must enable the developer exception page by calling `app.UseDeveloperExceptionPage` in `Configure`.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/error-handling.md:24]:
```suggestion
* App created with the current templates, that is, using [WebApplication.CreateBuilder](/dotnet/api/microsoft.aspnetcore.builder.webapplication.createbuilder).  Apps created using the [`WebHost.CreateDefaultBuilder`](xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder) must enable the developer exception page by calling `app.UseDeveloperExceptionPage` in `Configure`.
```

**Comment by DamianEdwards** [comment]
> @DamianEdwards @davidfowl @Tratcher Do you have any opinions on whether we should start enabling the developer exception page by default in development with the generic web host? It'd be easier to explain.

I'm inclined to just leave it be, i.e. not change it. The new API enables it, the existing API doesn't it, I think that's easy enough to understand.

**Pull request merged**